File: examples/demo_client.cpp
/*
 * Copyright 2017 DF Automation & Robotics Sdn. Bhd.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <bitset>
#include <ctime>
#include <iostream>
#include <zalpha_api/zalpha.hpp>

#ifdef _WINDOWS
#include <windows.h>
#else
#include <unistd.h>
#define Sleep(x) usleep((x)*1000)
#endif


const char* info =
  "------------------------------------------------------------------------------\n"
  "This demo client can perform the following operations:\n"
  "------------------------------------------------------------------------------\n"
  "1. Read version info.\n"
  "2. Set acceleration and deceleration to 2.0 m/s2.\n"
  "3. Set acceleration and deceleration to 0.1 m/s2.\n"
  "4. Read acceleration values.\n"
  "5. Move agv forward at 0.5 m/s for 3 seconds and stop.\n"
  "6. Move agv backward at 0.3 m/s for 3 seconds and stop.\n"
  "7. Rotate agv clockwise at 90 deg/s for 3 seconds and stop.\n"
  "8. Rotate agv anticlockwise at 60 deg/s for 3 seconds and stop.\n"
  "9. Forward agv in a straight movement to reach 2 m.\n"
  "10. Reverse agv in a straight movement to reach -1.5 m.\n"
  "11. Forward agv in a bezier movement to reach 2 m (front) and -1 m (right).\n"
  "12. Reverse agv in a bezier movement to reach -2 m (rear) and 1 m (left).\n"
  "13. Rotate agv clockwise for 90 deg.\n"
  "14. Rotate agv anticlockwise for 180 deg.\n"
  "15. Get action status.\n"
  "16. Pause action.\n"
  "17. Resume action.\n"
  "18. Stop action.\n"
  "19. Reset encoder to zero.\n"
  "20. Read encoder distance.\n"
  "21. Read encoder count.\n"
  "22. Read safety flag.\n"
  "23. Read encoder distance and safety flag.\n"
  "24. Read encoder count and safety flag.\n"
  "25. Read battery percentage.\n"
  "26. Activate charging.\n"
  "27. Read charging status.\n"
  "28. Read inputs.\n"
  "29. Set all outputs to HIGH.\n"
  "30. Set all outputs to LOW.\n"
  "31. Read outputs.\n"
  "32. Exit.\n"
  "------";

const int option_count = 32;

int getOption()
{
  std::cout << info << std::endl;
  int option;
  do
  {
    option = -1;
    std::cout << "Please enter an option (1-32): ";
    std::cin >> option;
    std::cin.clear();
    std::cin.ignore(10000, '\n');
  }
  while (option < 1 || option > option_count);

  std::cout << "------" << std::endl;
  return option;
}

void versionInfo(zalpha_api::Zalpha& agv)
{
  std::cout << "API client library version: " << zalpha_api::VERSION << std::endl;

  std::string version;
  if (agv.versionInfo(version))
  {
    std::cout << "API server version: " << version << " (";
    std::cout << (zalpha_api::versionCompatible(version) ? "compatible" : "incompatible");
    std::cout << ")" << std::endl;
  }
  else
  {
    std::cerr << "Failed to obtain API server version: " << agv.getErrorMessage() << std::endl;
  }
}

void setAcceleration(zalpha_api::Zalpha& agv, float acc)
{
  if (agv.setAcceleration(acc, acc))
  {
    std::cout << "Successfully set acceleration to " << acc << " m/s2." << std::endl;
  }
  else
  {
    std::cerr << "Failed to set acceleration: " << agv.getErrorMessage() << std::endl;
  }
}

void getAcceleration(zalpha_api::Zalpha& agv)
{
  float acc, dec;
  if (agv.getAcceleration(acc, dec))
  {
    std::cout << "Acceleration: " << acc << " m/s2" << std::endl;
    std::cout << "Deceleration: " << dec << " m/s2" << std::endl;
  }
  else
  {
    std::cerr << "Failed to read acceleration values: " << agv.getErrorMessage() << std::endl;
  }
}

void setSpeed(zalpha_api::Zalpha& agv, float left_speed, float right_speed, int duration)
{
  if (agv.setTargetSpeed(left_speed, right_speed))
  {
    std::cout << "Successfully set target speed to (" << left_speed << ", " << right_speed << ") m/s." << std::endl;
  }
  else
  {
    std::cerr << "Failed to set target speed: " << agv.getErrorMessage() << std::endl;
    return;
  }

  std::cout << "Waiting for " << duration << " seconds before stopping." << std::endl;
  time_t stop_time = time(NULL) + duration;
  while (time(NULL) < stop_time)
  {
    agv.setTargetSpeed(left_speed, right_speed);
    Sleep(10);
  }

  if (agv.setTargetSpeed(0.0, 0.0))
  {
    std::cout << "Successfully stop AGV." << std::endl;
  }
  else
  {
    std::cerr << "Failed to stop AGV: " << agv.getErrorMessage() << std::endl;
  }
}

void moveStraight(zalpha_api::Zalpha& agv, float speed, float distance, uint8_t laser_area)
{
  if (agv.moveStraight(speed, distance, laser_area))
  {
    std::cout << "Successfully start move_straight action." << std::endl;
  }
  else
  {
    std::cerr << "Failed to start move_straight action: " << agv.getErrorMessage() << std::endl;
  }
}

void moveBezier(zalpha_api::Zalpha& agv, float speed, float x, float y, float cp1_x, float cp1_y, float cp2_x, float cp2_y, uint8_t laser_area)
{
  if (agv.moveBezier(speed, x, y, cp1_x, cp1_y, cp2_x, cp2_y, laser_area))
  {
    std::cout << "Successfully start move_bezier action." << std::endl;
  }
  else
  {
    std::cerr << "Failed to start move_bezier action: " << agv.getErrorMessage() << std::endl;
  }
}

void rotate(zalpha_api::Zalpha& agv, float speed, float angle, uint8_t laser_area)
{
  if (agv.rotate(speed, angle, laser_area))
  {
    std::cout << "Successfully start rotate action." << std::endl;
  }
  else
  {
    std::cerr << "Failed to start rotate action: " << agv.getErrorMessage() << std::endl;
  }
}

void getActionStatus(zalpha_api::Zalpha& agv)
{
  uint8_t status;
  if (agv.getActionStatus(status))
  {
    if (status == agv.AC_COMPLETED)
    {
      std::cout << "Action status: Completed" << std::endl;
    }
    else if (status == agv.AC_IN_PROGRESS)
    {
      std::cout << "Action status: In progress" << std::endl;
    }
    else if (status == agv.AC_PAUSED)
    {
      std::cout << "Action status: Paused" << std::endl;
    }
    else if (status == agv.AC_SAFETY_TRIGGERED)
    {
      std::cout << "Action status: Safety triggered" << std::endl;
    }
    else
    {
      std::cout << "Action status: Unknown" << std::endl;
    }
  }
  else
  {
    std::cerr << "Failed to get action status: " << agv.getErrorMessage() << std::endl;
  }
}

void pauseAction(zalpha_api::Zalpha& agv)
{
  if (agv.pauseAction())
  {
    std::cout << "Successfully pause action." << std::endl;
  }
  else
  {
    std::cerr << "Failed to pause action: " << agv.getErrorMessage() << std::endl;
  }
}

void resumeAction(zalpha_api::Zalpha& agv)
{
  if (agv.resumeAction())
  {
    std::cout << "Successfully resume action." << std::endl;
  }
  else
  {
    std::cerr << "Failed to resume action: " << agv.getErrorMessage() << std::endl;
  }
}

void stopAction(zalpha_api::Zalpha& agv)
{
  if (agv.stopAction())
  {
    std::cout << "Successfully stop action." << std::endl;
  }
  else
  {
    std::cerr << "Failed to stop action: " << agv.getErrorMessage() << std::endl;
  }
}

void resetEncoder(zalpha_api::Zalpha& agv)
{
  if (agv.resetEncoder())
  {
    std::cout << "Successfully reset encoder to zero." << std::endl;
  }
  else
  {
    std::cerr << "Failed to reset encoder: " << agv.getErrorMessage() << std::endl;
  }
}

void getEncoder(zalpha_api::Zalpha& agv)
{
  double left_distance, right_distance;
  if (agv.getEncoder(left_distance, right_distance))
  {
    std::cout << "Encoder distance: (" << left_distance << ", " << right_distance << ") meters." << std::endl;
  }
  else
  {
    std::cerr << "Failed to read encoder distance: " << agv.getErrorMessage() << std::endl;
  }
}

void getRawEncoder(zalpha_api::Zalpha& agv)
{
  int64_t left_count, right_count;
  if (agv.getRawEncoder(left_count, right_count))
  {
    std::cout << "Encoder count: (" << left_count << ", " << right_count << ") pulses." << std::endl;
  }
  else
  {
    std::cerr << "Failed to read encoder count: " << agv.getErrorMessage() << std::endl;
  }
}

void printSafetyFlag(uint16_t safety_flag)
{
  std::cout << "Safety flags (Critical):" << std::endl;
  std::cout << " - Bumper front: " << (bool)(safety_flag & zalpha_api::Zalpha::SF_BUMPER_FRONT) << std::endl;
  std::cout << " - Bumper rear: " << (bool)(safety_flag & zalpha_api::Zalpha::SF_BUMPER_REAR) << std::endl;
  std::cout << " - Emergency button: " << (bool)(safety_flag & zalpha_api::Zalpha::SF_EMERGENCY_BUTTON) << std::endl;
  std::cout << " - External input: " << (bool)(safety_flag & zalpha_api::Zalpha::SF_EXTERNAL_INPUT) << std::endl;
  std::cout << " - Motor fault: " << (bool)(safety_flag & zalpha_api::Zalpha::SF_MOTOR_FAULT) << std::endl;
  std::cout << " - Wheel slippage: " << (bool)(safety_flag & zalpha_api::Zalpha::SF_WHEEL_SLIPPAGE) << std::endl;
  std::cout << " - Charger connected: " << (bool)(safety_flag & zalpha_api::Zalpha::SF_CHARGER_CONNECTED) << std::endl;
  std::cout << "Safety flags (Non-critical):" << std::endl;
  std::cout << " - Laser far area blocked: " << (bool)(safety_flag & zalpha_api::Zalpha::SF_LASER_FAR_BLOCKED) << std::endl;
  std::cout << " - Laser middle area blocked: " << (bool)(safety_flag & zalpha_api::Zalpha::SF_LASER_MIDDLE_BLOCKED) << std::endl;
  std::cout << " - Laser near area blocked: " << (bool)(safety_flag & zalpha_api::Zalpha::SF_LASER_NEAR_BLOCKED) << std::endl;
  std::cout << " - Laser malfunction: " << (bool)(safety_flag & zalpha_api::Zalpha::SF_LASER_MALFUNCTION) << std::endl;
}

void getSafetyFlag(zalpha_api::Zalpha& agv)
{
  uint16_t safety_flag;
  if (agv.getSafetyFlag(safety_flag))
  {
    printSafetyFlag(safety_flag);
  }
  else
  {
    std::cerr << "Failed to read safety flag: " << agv.getErrorMessage() << std::endl;
  }
}

void getEncoderAndSafetyFlag(zalpha_api::Zalpha& agv)
{
  double left_distance, right_distance;
  uint16_t safety_flag;
  if (agv.getEncoderAndSafetyFlag(left_distance, right_distance, safety_flag))
  {
    std::cout << "Encoder distance: (" << left_distance << ", " << right_distance << ") meters." << std::endl;
    printSafetyFlag(safety_flag);
  }
  else
  {
    std::cerr << "Failed to read encoder distance and safety flag: " << agv.getErrorMessage() << std::endl;
  }
}

void getRawEncoderAndSafetyFlag(zalpha_api::Zalpha& agv)
{
  int64_t left_count, right_count;
  uint16_t safety_flag;
  if (agv.getRawEncoderAndSafetyFlag(left_count, right_count, safety_flag))
  {
    std::cout << "Encoder count: (" << left_count << ", " << right_count << ") pulses." << std::endl;
    printSafetyFlag(safety_flag);
  }
  else
  {
    std::cerr << "Failed to read encoder count and safety flag: " << agv.getErrorMessage() << std::endl;
  }
}

void getBattery(zalpha_api::Zalpha& agv)
{
  float battery;
  if (agv.getBattery(battery))
  {
    std::cout << "Battery percentage: " << battery << std::endl;
  }
  else
  {
    std::cerr << "Failed to read battery percentage: " << agv.getErrorMessage() << std::endl;
  }
}

void setCharging(zalpha_api::Zalpha& agv, bool enable)
{
  const char* action = (enable) ? "activate" : "deactivate";

  if (agv.setCharging(enable))
  {
    std::cout << "Successfully " << action << " charging." << std::endl;
  }
  else
  {
    std::cerr << "Failed to " << action << " charging: " << agv.getErrorMessage() << std::endl;
  }
}

void getCharging(zalpha_api::Zalpha& agv)
{
  uint8_t state;
  if (agv.getCharging(state))
  {
    std::cout << "Charging status:" << std::endl;
    std::cout << " - Plug type: ";
    std::cout << (state & zalpha_api::Zalpha::CH_AUTO_MANUAL ? "Manual" : "Auto") << std::endl;
    std::cout << " - Charging: " << (bool)(state & zalpha_api::Zalpha::CH_CHARGING) << std::endl;
    std::cout << " - Battery full: " << (bool)(state & zalpha_api::Zalpha::CH_BATTERY_FULL) << std::endl;
  }
  else
  {
    std::cerr << "Failed to read charging status: " << agv.getErrorMessage() << std::endl;
  }
}

void printBits(uint32_t bits)
{
  std::cout << std::bitset<8>(bits >> 24) << " ";
  std::cout << std::bitset<8>(bits >> 16) << " ";
  std::cout << std::bitset<8>(bits >> 8) << " ";
  std::cout << std::bitset<8>(bits);
}

void getInputs(zalpha_api::Zalpha& agv)
{
  uint32_t inputs;
  if (agv.getInputs(inputs))
  {
    std::cout << "Inputs: (MSB) ";
    printBits(inputs);
    std::cout << " (LSB)" << std::endl;
  }
  else
  {
    std::cerr << "Failed to read inputs: " << agv.getErrorMessage() << std::endl;
  }
}

void setOutputs(zalpha_api::Zalpha& agv, uint32_t outputs, uint32_t mask)
{
  if (agv.setOutputs(outputs, mask))
  {
    std::cout << "Successfully set outputs." << std::endl;
  }
  else
  {
    std::cerr << "Failed to set outputs: " << agv.getErrorMessage() << std::endl;
  }
}

void getOutputs(zalpha_api::Zalpha& agv)
{
  uint32_t outputs;
  if (agv.getOutputs(outputs))
  {
    std::cout << "Outputs: (MSB) ";
    printBits(outputs);
    std::cout << " (LSB)" << std::endl;
  }
  else
  {
    std::cerr << "Failed to read outputs: " << agv.getErrorMessage() << std::endl;
  }
}


void execute(zalpha_api::Zalpha& agv, int option)
{
  switch (option)
  {
  case 1:
    versionInfo(agv);
    break;
  case 2:
    setAcceleration(agv, 2.0f);
    break;
  case 3:
    setAcceleration(agv, 0.1f);
    break;
  case 4:
    getAcceleration(agv);
    break;
  case 5:
    setSpeed(agv, 0.5f, 0.5f, 3);
    break;
  case 6:
    setSpeed(agv, -0.3f, -0.3f, 3);
    break;
  case 7:
  {
    const float baseWidth = 0.51f;
    float linearSpeed = 90 / 180.0 * 3.14159 * baseWidth / 2.0;
    setSpeed(agv, linearSpeed, -linearSpeed, 3);
    break;
  }
  case 8:
  {
    const float baseWidth = 0.51f;
    float linearSpeed = 60 / 180.0 * 3.14159 * baseWidth / 2.0;
    setSpeed(agv, -linearSpeed, linearSpeed, 3);
    break;
  }
  case 9:
  {
    float distance = 2.0f;
    moveStraight(agv, 0.6f, distance, 1);
    break;
  }
  case 10:
  {
    float distance = -1.5f;
    moveStraight(agv, 0.6f, distance, 1);
    break;
  }
  case 11:
  {
    float end_x = 2.0f, end_y = -1.0f;
    float cp1_x = 1.0f, cp1_y = 0.0f;
    float cp2_x = 1.0f, cp2_y = -1.0f;
    moveBezier(agv, 0.4, end_x, end_y, cp1_x, cp1_y, cp2_x, cp2_y, 1);
    break;
  }
  case 12:
  {
    float end_x = -2.0f, end_y = 1.0f;
    float cp1_x = -1.0f, cp1_y = 0.0f;
    float cp2_x = -1.0f, cp2_y = 1.0f;
    moveBezier(agv, 0.4, end_x, end_y, cp1_x, cp1_y, cp2_x, cp2_y, 1);
    break;
  }
  case 13:
  {
    float angle = -90 / 180.0 * 3.14159;
    rotate(agv, 0.4, angle, 1);
    break;
  }
  case 14:
  {
    float angle = 180 / 180.0 * 3.14159;
    rotate(agv, 0.4, angle, 1);
    break;
  }
  case 15:
    getActionStatus(agv);
    break;
  case 16:
    pauseAction(agv);
    break;
  case 17:
    resumeAction(agv);
    break;
  case 18:
    stopAction(agv);
    break;
  case 19:
    resetEncoder(agv);
    break;
  case 20:
    getEncoder(agv);
    break;
  case 21:
    getRawEncoder(agv);
    break;
  case 22:
    getSafetyFlag(agv);
    break;
  case 23:
    getEncoderAndSafetyFlag(agv);
    break;
  case 24:
    getRawEncoderAndSafetyFlag(agv);
    break;
  case 25:
    getBattery(agv);
    break;
  case 26:
    setCharging(agv, true);
    break;
  case 27:
    getCharging(agv);
    break;
  case 28:
    getInputs(agv);
    break;
  case 29:
    setOutputs(agv, 0xffff, 0xffff);
    break;
  case 30:
    setOutputs(agv, 0x0000, 0xffff);
    break;
  case 31:
    getOutputs(agv);
    break;
  }

  std::cout << std::endl << "Press enter to continue...";
  std::cin.ignore(10000, '\n');
  std::cout << std::endl << std::endl;
}

int main(int argc, char** argv)
{
  if (argc != 2)
  {
    std::cout << "Usage: demo_client <server_ip_address>" << std::endl;
    return 0;
  }

  zalpha_api::Zalpha agv;
  if (!agv.connect(argv[1]))
  {
    std::cerr << "Error connecting to API server: " << agv.getErrorMessage() << std::endl;
    return 1;
  }

  while (true)
  {
    int option = getOption();
    if (option == option_count) break;
    execute(agv, option);
  }

  std::cout << "Bye!" << std::endl;
  return 0;
}

File: python/examples/demo_client.py
# -*- coding: utf-8 -*-
# Copyright 2017 DF Automation & Robotics Sdn. Bhd.  All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import unicode_literals

import sys
import time
import zalpha_api


if sys.version_info[0] < 3:
    user_input = raw_input
else:
    user_input = input


info = """
------------------------------------------------------------------------------
This demo client can perform the following operations:
------------------------------------------------------------------------------
1. Read version info.
2. Set acceleration and deceleration to 2.0 m/s2.
3. Set acceleration and deceleration to 0.1 m/s2.
4. Read acceleration values.
5. Move agv forward at 0.5 m/s for 3 seconds and stop.
6. Move agv backward at 0.3 m/s for 3 seconds and stop.
7. Rotate agv clockwise at 90 deg/s for 3 seconds and stop.
8. Rotate agv anticlockwise at 60 deg/s for 3 seconds and stop.
9. Forward agv in a straight movement to reach 2 m.
10. Reverse agv in a straight movement to reach -1.5 m.
11. Forward agv in a bezier movement to reach 2 m (front) and -1 m (right).
12. Reverse agv in a bezier movement to reach -2 m (rear) and 1 m (left).
13. Rotate agv clockwise for 90 deg.
14. Rotate agv anticlockwise for 180 deg.
15. Get action status.
16. Pause action.
17. Resume action.
18. Stop action.
19. Reset encoder to zero.
20. Read encoder distance.
21. Read encoder count.
22. Read safety flag.
23. Read encoder distance and safety flag.
24. Read encoder count and safety flag.
25. Read battery percentage.
26. Activate charging.
27. Read charging status.
28. Read inputs.
29. Set all outputs to HIGH.
30. Set all outputs to LOW.
31. Read outputs.
32. Exit.
------"""

option_count = 32


def get_option():
    print(info)
    while True:
        option = user_input('Please enter an option (1-32): ')
        try:
            option = int(option)
        except:
            pass
        if isinstance(option, int) and option >= 1 and option <= option_count:
            break
    print('------')
    return option


def version_info(agv):
    print('API client library version: %s' % zalpha_api.__version__)

    try:
        version = agv.version_info()
        print('API server version: %s (%s)' % (version,
            'compatible' if zalpha_api.version_compatible(version) else
            'incompatible'))
    except Exception as ex:
        print('Failed to obtain API server version: %s' % ex)


def set_acceleration(agv, acc):
    try:
        agv.set_acceleration(acc, acc)
        print('Successfully set acceleration to %s m/s2.' % acc)
    except Exception as ex:
        print('Failed to set acceleration: %s' % ex)


def get_acceleration(agv):
    try:
        acc, dec = agv.get_acceleration()
        print('Acceleration: %s m/s2' % acc)
        print('Deceleration: %s m/s2' % dec)
    except Exception as ex:
        print('Failed to read acceleration values: %s' % ex)


def set_speed(agv, left_speed, right_speed, duration):
    try:
        agv.set_target_speed(left_speed, right_speed)
        print('Successfully set target speed to (%s, %s) m/s.' % (
            left_speed, right_speed))
    except Exception as ex:
        print('Failed to set target speed: %s' % ex)
        return

    print('Waiting for %d seconds before stopping.' % duration)
    stop_time = time.time() + duration
    while time.time() < stop_time:
        try:
            agv.set_target_speed(left_speed, right_speed)
        except:
            pass
        time.sleep(0.01)

    try:
        agv.set_target_speed(0.0, 0.0)
        print('Successfully stop AGV.')
    except Exception as ex:
        print('Failed to stop AGV: %s' % ex)


def move_straight(agv, speed, distance, laser_area):
    try:
        agv.move_straight(speed, distance, laser_area)
        print('Successfully start move_straight action.')
    except Exception as ex:
        print('Failed to start move_straight action: %s' % ex)


def move_bezier(agv, speed, x, y, cp1_x, cp1_y, cp2_x, cp2_y, laser_area):
    try:
        agv.move_bezier(speed, x, y, cp1_x, cp1_y, cp2_x, cp2_y, laser_area)
        print('Successfully start move_bezier action.')
    except Exception as ex:
        print('Failed to start move_bezier action: %s' % ex)


def rotate(agv, speed, angle, laser_area):
    try:
        agv.rotate(speed, angle, laser_area)
        print('Successfully start rotate action.')
    except Exception as ex:
        print('Failed to start rotate action: %s' % ex)


def get_action_status(agv):
    try:
        status = agv.get_action_status()
        if status == agv.AC_COMPLETED:
            print('Action status: Completed')
        elif status == agv.AC_IN_PROGRESS:
            print('Action status: In progress')
        elif status == agv.AC_PAUSED:
            print('Action status: Paused')
        elif status == agv.AC_SAFETY_TRIGGERED:
            print('Action status: Safety triggered')
        else:
            print('Action status: Unknown')
    except Exception as ex:
        print('Failed to get action status: %s' % ex)


def pause_action(agv):
    try:
        agv.pause_action()
        print('Successfully pause action.')
    except Exception as ex:
        print('Failed to pause action: %s' % ex)


def resume_action(agv):
    try:
        agv.resume_action()
        print('Successfully resume action.')
    except Exception as ex:
        print('Failed to resume action: %s' % ex)


def stop_action(agv):
    try:
        agv.stop_action()
        print('Successfully stop action.')
    except Exception as ex:
        print('Failed to stop action: %s' % ex)


def reset_encoder(agv):
    try:
        agv.reset_encoder()
        print('Successfully reset encoder to zero.')
    except Exception as ex:
        print('Failed to reset encoder: %s' % ex)


def get_encoder(agv):
    try:
        left_distance, right_distance = agv.get_encoder()
        print('Encoder distance: (%s, %s) meters.' % (left_distance, right_distance))
    except Exception as ex:
        print('Failed to read encoder distance: %s' % ex)


def get_raw_encoder(agv):
    try:
        left_count, right_count = agv.get_raw_encoder()
        print('Encoder count: (%s, %s) pulses.' % (left_count, right_count))
    except Exception as ex:
        print('Failed to read encoder count: %s' % ex)


def print_safety_flag(safety_flag):
    print('Safety flags (Critical):')
    print(' - Bumper front: %s' % bool(safety_flag & zalpha_api.Zalpha.SF_BUMPER_FRONT))
    print(' - Bumper rear: %s' % bool(safety_flag & zalpha_api.Zalpha.SF_BUMPER_REAR))
    print(' - Emergency button: %s' % bool(safety_flag & zalpha_api.Zalpha.SF_EMERGENCY_BUTTON))
    print(' - External input: %s' % bool(safety_flag & zalpha_api.Zalpha.SF_EXTERNAL_INPUT))
    print(' - Motor fault: %s' % bool(safety_flag & zalpha_api.Zalpha.SF_MOTOR_FAULT))
    print(' - Wheel slippage: %s' % bool(safety_flag & zalpha_api.Zalpha.SF_WHEEL_SLIPPAGE))
    print(' - Charger connected: %s' % bool(safety_flag & zalpha_api.Zalpha.SF_CHARGER_CONNECTED))
    print('Safety flags (Non-critical):')
    print(' - Laser far area blocked: %s' % bool(safety_flag & zalpha_api.Zalpha.SF_LASER_FAR_BLOCKED))
    print(' - Laser middle area blocked: %s' % bool(safety_flag & zalpha_api.Zalpha.SF_LASER_MIDDLE_BLOCKED))
    print(' - Laser near area blocked: %s' % bool(safety_flag & zalpha_api.Zalpha.SF_LASER_NEAR_BLOCKED))
    print(' - Laser malfunction: %s' % bool(safety_flag & zalpha_api.Zalpha.SF_LASER_MALFUNCTION))


def get_safety_flag(agv):
    try:
        safety_flag = agv.get_safety_flag()
        print_safety_flag(safety_flag)
    except Exception as ex:
        print('Failed to read safety flag: %s' % ex)


def get_encoder_and_safety_flag(agv):
    try:
        left_distance, right_distance, safety_flag = agv.get_encoder_and_safety_flag()
        print('Encoder distance: (%s, %s) meters.' % (left_distance, right_distance))
        print_safety_flag(safety_flag)
    except Exception as ex:
        print('Failed to read encoder distance and safety flag: %s' % ex)


def get_raw_encoder_and_safety_flag(agv):
    try:
        left_count, right_count, safety_flag = agv.get_raw_encoder_and_safety_flag()
        print('Encoder count: (%s, %s) pulses.' % (left_count, right_count))
        print_safety_flag(safety_flag)
    except Exception as ex:
        print('Failed to read encoder count and safety flag: %s' % ex)


def get_battery(agv):
    try:
        battery = agv.get_battery()
        print('Battery percentage: %s' % battery)
    except Exception as ex:
        print('Failed to read battery percentage: %s' % ex)


def set_charging(agv, enable):
    action = 'activate' if enable else 'deactivate'
    try:
        agv.set_charging(enable)
        print('Successfully %s charging.' % action)
    except Exception as ex:
        print('Failed to %s charging: %s' % (action, ex))


def get_charging(agv):
    try:
        state = agv.get_charging()
        print('Charging status:')
        print(' - Plug type: %s' % ('Manual' if state & zalpha_api.Zalpha.CH_AUTO_MANUAL else 'Auto'))
        print(' - Charging: %s' % bool(state & zalpha_api.Zalpha.CH_CHARGING))
        print(' - Battery full: %s' % bool(state & zalpha_api.Zalpha.CH_BATTERY_FULL))
    except Exception as ex:
        print('Failed to read charging status: %s' % ex)


def format_bits(bits):
    return '{0:08b} {1:08b} {2:08b} {3:08b}'.format(
        bits >> 24 & 0xff,
        bits >> 16 & 0xff,
        bits >> 8 & 0xff,
        bits & 0xff)


def get_inputs(agv):
    try:
        inputs = agv.get_inputs()
        print('Inputs: (MSB) %s (LSB)' % format_bits(inputs))
    except Exception as ex:
        print('Failed to read inputs: %s' % ex)


def set_outputs(agv, outputs, mask):
    try:
        agv.set_outputs(outputs, mask)
        print('Successfully set outputs.')
    except Exception as ex:
        print('Failed to set outputs: %s' % ex)


def get_outputs(agv):
    try:
        outputs = agv.get_outputs()
        print('Outputs: (MSB) %s (LSB)' % format_bits(outputs))
    except Exception as ex:
        print('Failed to read outputs: %s' % ex)


def execute(agv, option):
    if option == 1:
        version_info(agv)

    elif option == 2:
        set_acceleration(agv, 2.0)

    elif option == 3:
        set_acceleration(agv, 0.1)

    elif option == 4:
        get_acceleration(agv)

    elif option == 5:
        set_speed(agv, 0.5, 0.5, 3)

    elif option == 6:
        set_speed(agv, -0.3, -0.3, 3)

    elif option == 7:
        base_width = 0.51
        linear_speed = 90.0 / 180.0 * 3.14159 * base_width / 2.0
        set_speed(agv, linear_speed, -linear_speed, 3)

    elif option == 8:
        base_width = 0.51
        linear_speed = 60.0 / 180.0 * 3.14159 * base_width / 2.0
        set_speed(agv, -linear_speed, linear_speed, 3)

    elif option == 9:
        distance = 2.0
        move_straight(agv, 0.6, distance, 1)

    elif option == 10:
        distance = -1.5
        move_straight(agv, 0.6, distance, 1)

    elif option == 11:
        end = (2, -1)
        cp1 = (1, 0)
        cp2 = (1, -1)
        move_bezier(agv, 0.4, end[0], end[1], cp1[0], cp1[1], cp2[0], cp2[1], 1)

    elif option == 12:
        end = (-2, 1)
        cp1 = (-1, 0)
        cp2 = (-1, 1)
        move_bezier(agv, 0.4, end[0], end[1], cp1[0], cp1[1], cp2[0], cp2[1], 1)

    elif option == 13:
        angle = -90.0 / 180.0 * 3.14159
        rotate(agv, 0.4, angle, 1)

    elif option == 14:
        angle = 180.0 / 180.0 * 3.14159
        rotate(agv, 0.4, angle, 1)

    elif option == 15:
        get_action_status(agv)

    elif option == 16:
        pause_action(agv)

    elif option == 17:
        resume_action(agv)

    elif option == 18:
        stop_action(agv)

    elif option == 19:
        reset_encoder(agv)

    elif option == 20:
        get_encoder(agv)

    elif option == 21:
        get_raw_encoder(agv)

    elif option == 22:
        get_safety_flag(agv)

    elif option == 23:
        get_encoder_and_safety_flag(agv)

    elif option == 24:
        get_raw_encoder_and_safety_flag(agv)

    elif option == 25:
        get_battery(agv)

    elif option == 26:
        set_charging(agv, True)

    elif option == 27:
        get_charging(agv)

    elif option == 28:
        get_inputs(agv)

    elif option == 29:
        set_outputs(agv, 0xffff, 0xffff)

    elif option == 30:
        set_outputs(agv, 0x0000, 0xffff)

    elif option == 31:
        get_outputs(agv)

    user_input('\nPress enter to continue...')
    print('\n')


if __name__ == '__main__':
    if len(sys.argv) != 2:
        print('Usage: demo_client.py <server_ip_address>')
        sys.exit(0)

    agv = zalpha_api.Zalpha()
    try:
        agv.connect(sys.argv[1])
    except Exception as ex:
        print('Error connecting to API server: %s' % ex)
        sys.exit(1)

    while True:
        option = get_option()
        if option == option_count:
            break
        execute(agv, option)

    print('Bye!')
